### DISCUSSION TITLE
Does qlora adapter only support BitsAndBytes quantization model ?

### DISCUSSION THREAD

**Discussion by tarukumar** [vllm-project/vllm#9544] [Q&A]
What i have observed is when I try to deploy the model using `qlora_adapter_name_or_path` for qlora adapter it fails to deploy with the error mentioned in the line https://github.com/vllm-project/vllm/blob/main/vllm/engine/arg_utils.py#L899-L911. The question is to deploy qlora adpater should i use `--lora-modules` or `adapter-cache` parameter for qlora adapter? What is the best approach here ?

**Comment by jeejeelee** [comment]
See: https://github.com/vllm-project/vllm/blob/main/examples/lora_with_quantization_inference.py

**Reply by tarukumar** [reply]
Thanks for the response. I have looked into the code and still this doesn't  clear my concern with the parameter `qlora_adapter_name_or_path` is only applicable for BnB quant model from where my original equation arised

Also we have open issue https://github.com/vllm-project/vllm/issues/3225

**Reply by jeejeelee** [reply]
Yes, `qlora_adapter_name_or_path` is only applicable for BnB quant model.

**Reply by tarukumar** [reply]
Okay so for other qunat model to load qlora adapter we need to use `lora-modules` argument right? Also, my main question is why we can not use  qlora_adapter_name_or_path argument for other quant model. What is the technical bottleneck beacsue this doesn’t seem consistent for one specific quant method we have to use specific argument and for others we have to use different

**Reply by jeejeelee** [reply]
vllm first added support for LoRA with other quantization models (such as GPTQ, awq, etc.) before supporting QLoRA. I think this is due to implementation logic reasons. Would you please help explain this, @chenqianfzh, Thank you.

**Reply by chenqianfzh** [reply]
@jeejeelee sure, my pleasure.

QLoRA and 4-bit quantization were originally introduced together in the paper "QLoRA: Efficient Finetuning of Quantized LLMs" (https://arxiv.org/abs/2305.14314). In practice, most QLoRA models utilize bitsandbytes for 4-bit quantization.

However, as @tarukumar observed from the code, QLoRA adapters are not restricted to supporting bitsandbytes quantization exclusively. We are prepared to modify the code if other quantization models require such functionality.

Hope this helps. :-)

**Reply by tarukumar** [reply]
@chenqianfzh That would be good if we can expand that. Thank you!

**Reply by tarukumar** [reply]
@chenqianfzh should i open the issue for this support ?

**Reply by chenqianfzh** [reply]
> @chenqianfzh should i open the issue for this support ?

Do you see many such use cases? I would add the feature if it is a popular demand.

**Reply by tarukumar** [reply]
It would be definitely useful may be we can add this as a part of https://github.com/bd-iaas-us/vllm/issues/2

**Comment by chenqianfzh** [comment]
@tarukumar Feel free to open an issue at https://github.com/bd-iaas-us/vllm/issues and assign it to me.

In this issue, please let us know why existing Quantization + LoRA  solution in vLLM does not suffice, as well as some models in need of this feature. 

Thanks!
